lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -1,98 +1,91 @@
 {-# LANGUAGE LambdaCase #-}
 
-module LambdaBuffers.Compiler.TypeClass.Solve (solveM, solve, Overlap (..)) where
-
-import LambdaBuffers.Compiler.TypeClass.Pat (
-  Pat (AppP, DecP, ProdP, RecP, RefP, SumP, VarP, (:*), (:=)),
-  matches,
- )
-import LambdaBuffers.Compiler.TypeClass.Rules (
-  Class (csupers),
-  Constraint (C),
-  Rule ((:<=)),
-  mapPat,
-  ruleHeadClass,
-  ruleHeadPat,
- )
+module LambdaBuffers.Compiler.TypeClass.Solve (solve, inst, Overlap (..)) where
 
 import Control.Monad.Except (throwError)
 import Control.Monad.Reader (ReaderT, runReaderT)
 import Control.Monad.Reader.Class (MonadReader (ask))
 import Control.Monad.Writer.Class (MonadWriter (tell))
 import Control.Monad.Writer.Strict (WriterT, execWriterT)
 import Data.Foldable (traverse_)
-import Data.List (foldl')
+import Data.Map qualified as M
 import Data.Set qualified as S
 import Data.Text (Text)
+import LambdaBuffers.Compiler.TypeClass.Pat (Exp (AppE, ConsE, DecE, LabelE, LitE, NilE, ProdE, RecE, RefE, SumE), ExpressionLike ((*:), (*=)), Literal (TyVar), Pat (AppP, ConsP, DecP, LabelP, LitP, NilP, ProdP, RecP, RefP, SumP, VarP), matches)
+import LambdaBuffers.Compiler.TypeClass.Rules (Class (csupers), Constraint (C), Rule ((:<=)), ruleClass, ruleHead)
 
 {- Pattern/Template/Unification variable  substitution.
    Given a string that represents a variable name,
    and a type to instantiate variables with that name to,
    performs the instantiation
 -}
-subV :: Text -> Pat -> Pat -> Pat
-subV varNm t = \case
-  var@(VarP v) -> if v == varNm then t else var
-  x :* xs -> subV varNm t x :* subV varNm t xs
-  l := x -> subV varNm t l := subV varNm t x
-  ProdP xs -> ProdP (subV varNm t xs)
-  RecP xs -> RecP (subV varNm t xs)
-  SumP xs -> SumP (subV varNm t xs)
-  AppP t1 t2 -> AppP (subV varNm t t1) (subV varNm t t2)
-  RefP n x -> RefP (subV varNm t n) (subV varNm t x)
-  DecP a b c -> DecP (subV varNm t a) (subV varNm t b) (subV varNm t c)
-  other -> other
+subV :: M.Map Text Exp -> Pat -> Exp
+subV cxt = \case
+  (VarP v) -> case M.lookup v cxt of
+    Nothing -> LitE (TyVar v)
+    Just t -> t
+  ConsP x xs -> subV cxt x *: subV cxt xs
+  LabelP l x -> subV cxt l *= subV cxt x
+  ProdP xs -> ProdE (subV cxt xs)
+  RecP xs -> RecE (subV cxt xs)
+  SumP xs -> SumE (subV cxt xs)
+  AppP t1 t2 -> AppE (subV cxt t1) (subV cxt t2)
+  RefP n x -> RefE (subV cxt n) (subV cxt x)
+  DecP a b c -> DecE (subV cxt a) (subV cxt b) (subV cxt c)
+  LitP l -> LitE l
+  NilP -> NilE
```

Comment:
Ok I think I get it, the cxt (rename it to ctx?) is established by the TyAbs from the enclosing TyDef. So here, it looks like you're literally doing lc substitution. 

I got confused because I would have expected to see a AbsP somewhere that introduces the arguments in the context.

I'm still confused by what this does to `DecP`?

Reply:
I should just remove `DecP`. I agree that this doesn't make much sense for that. (I probably mentioned this but `DecP` is mostly an artifact of me wanting to use Pat (or well Exp I guess) for codegen)

Reply:
btw it's not really "lc-substitution" it's just substitution, which was formalized (iirc) years before the LC existed. Anyway, marking this resolved